### DISCUSSION TITLE
Natural Text Editing: "command delete" should delete whole line, not one character

### DISCUSSION THREAD

**Discussion by itswil** [ghostty-org/ghostty#3232] [Q&A] [answered]
```
macOS 15.2

Version 1.0.0
Build 8469
Commit 4b4d4062
```

**Current behaviour**
Comparing to iTerm2's Natural Text Editing, Ghostty already has:
`option ←` = move back one word
`option →` = move forward one word
`option delete` = delete one word

`command ←` = move to start of line
`command →` = move to end of line

**Issue**
However, when pressing `command delete`, only one character is deleted.
On iTerm2, the whole line is deleted.

**Expected result**
To delete whole line

**Comment by marcrdez** [comment, marked as answer]
Try this in your ghostty config, it worked for me:

```
# Works in fish
keybind = cmd+backspace=text:\x15
# Works in zsh
keybind = cmd+backspace=esc:w
```

**Reply by colin-rocks** [reply]
almost, but I believe what most people are expecting by default is `keybind = cmd+backspace=esc:w` which deletes from the cursor backwards, whereas `text:\x15` deletes the entire line.

**Reply by david-crespo** [reply]
Heh, I have `text:\x15` in my config and didn't notice it did the whole line instead of starting from cursor. `esc:w` isn't working for me though. It causes the keybind to do nothing.

**Reply by marcrdez** [reply]
`text:\x15` its working from the cursor in my ghostty, maybe because i have fish behind, maybe it does not work in other shell idk. `esc:w` does nothing to me also.

**Reply by colin-rocks** [reply]
Interesting, I’m using zsh with oh-my-zsh; I have no special keybinds in my zshrc. afaict omz does not bind anything special or extra. esc:w works for me in and outside of tmux, too.

Far too ignorant to understand why but would be curious if someone has an explanation to share.

**Reply by xheisenbugx** [reply]
`keybind = cmd+backspace=esc:w ` works for me, i am using zsh, it deletes from your current position to the start of the line

**Reply by marcrdez** [reply]
Updated the answer with both options... I do not have anything weird in my fish that makes that work, so I assume it just shell differences....

**Reply by david-crespo** [reply]
Funny, I’m using zsh on mac, so it sounds like it should work. I will experiment some more.

**Reply by altano** [reply]
`keybind = cmd+backspace=esc:w` worked for me on zsh, thank you.

FWIW iTerm deletes the entire line, not from the cursor backwards, and that has always been super annoying. Most native text controls on macOS are cursor backwards, NOT entire line.

**Reply by nikitajz** [reply]
`text:\x15` sends Ctrl-U as mentioned in the ghostty docs as well
By default, it's not bound in zsh and requires explicitly specifying in the zsh config

So altogether:
Add to `~/.zshrc` (or alternative) config:
```
bindkey \^U backward-kill-line  # fix Ctrl-U: remove to the beginning of the line
```
Add to `ghostty` config:
```
keybind = cmd+backspace=text:\x15
```

**Comment by altano** [comment]
Interestingly, Terminal.app is even more poorly behaved. I tested a few text controls and here's what I saw:

| Application | Command-backspace supported | Option-backspace supported |
|-|-|-|
| Safari address bar | ✅ | ✅ |
| TextEdit | ✅ | ✅ |
| Finder search box | ✅ | ✅ |
| iTerm | ⚠️ [^iterm-clarification] | ✅ |
| ghostty | ❌ | ✅ |
| Terminal.app | ❌ | ❌ |

[^iterm-clarification]: always deletes the whole line, not from the cursor location to the start of the line

**Comment by luispadron** [comment]
Shout out to @altano for the breakdown and the `keybind = cmd+backspace=esc:w` config to match most macOS behavior. 

IMO this should be the default on macOS (not sure about other platforms). Immediately found it confusing that it deletes a single character, easy enough to configure it but unintuitive.

**Comment by raymondji** [comment]
To add another data point: when I press `command delete`, nothing happens at all. `keybind = cmd+backspace=esc:w` did not work for me either (again nothing happens).

I'm using zsh on macOS without oh-my-zsh or much of anything installed.

**Reply by altano** [reply]
Did you remember to reload your configuration after saving the change?

**Reply by raymondji** [reply]
Yes! Have tried runtime reloading and also restarting the terminal.

**Reply by david-crespo** [reply]
Same. I also tried commenting out some stuff in my zshrc setting up emacs keybinds and a couple other shortcuts.

**Reply by itswil** [reply]
+1; looks like it's the accepted answer though. May have to use `keybind = cmd+backspace=text:\x15` as a temporary workaround.
(Using zsh on macOS, without oh-my-zsh; tried with blank `.zshrc` too)

**Reply by raymondji** [reply]
`keybind = cmd+backspace=text:\x15` worked for me, thank you!

**Comment by mitchellh** [comment]
Approved. #3646

**Reply by marcrdez** [reply]
@mitchellh is it any way that can be apply depending on the shell? `esc:w` works in certain zsh, but `text:\x15` works perfectly with bash/fish.

**Comment by rrotter** [comment]
`keybind = cmd+backspace=text:\x15` does what's expected in bash (delete to the left of the cursor) but with default zsh keybindings it would [kill-whole-line](https://zsh.sourceforge.io/Doc/Release/Zsh-Line-Editor.html#index-kill_002dwhole_002dline), which isn't what we want. The ideal operation is [backward-kill-line](https://zsh.sourceforge.io/Doc/Release/Zsh-Line-Editor.html#index-backward_002dkill_002dline), which is, annoyingly, unbound by default. [edit: To get the bash/fish behavior for `\x15` in zsh you need to add `bindkey '^U' backward-kill-line` to `.zshrc`.]

esc+w is [copy-region-as-kill](https://zsh.sourceforge.io/Doc/Release/Zsh-Line-Editor.html#index-copy_002dregion_002das_002dkill). I've never used this and can't get it to work. Looking at this doc, it appears that it requires a "mark" that would normally be applied interactively. Maybe there is some default feature in zsh that I and others that can't get the `keybind = cmd+backspace=esc:w` to work have disabled?

**Reply by itswil** [reply]
Thanks for this explanation, I wouldn't know where to start looking otherwise.

So I believe that the `zsh` users who say `keybind = cmd+backspace=esc:w` works correctly might all be using Oh My Zsh.

Partially related to this [line](https://github.com/ohmyzsh/ohmyzsh/blob/f733dc340b2a1c5b2e61a4da7de790b2f557175f/lib/key-bindings.zsh#L114):
```
bindkey '\ew' kill-region
```
I haven't found the exact line that causes it since [kill-region](https://zsh.sourceforge.io/Doc/Release/Zsh-Line-Editor.html#index-kill_002dregion) is not the correct action on it's own, but the `esc:w` fix ceases to work without it.

**Reply by david-crespo** [reply]
Confirmed, adding 

```sh
bindkey '\ew' backward-kill-line
```

to my zshrc makes 

```
keybind = cmd+backspace=esc:w
```

work!

**Comment by AlexXi19** [comment]
Option + backspace & Cmd + backspace also does not work in zsh vi mode (`keybind = cmd+backspace=text:\x15` works though). Option + backspace also does not delete word in neovim.

**Reply by AlexXi19** [reply]
Might not be the best solution, but i was able to get it working as close to my previous iterm setup with the following: 

ghostty config
```
keybind = cmd+backspace=esc:w
keybind = cmd+backspace=text:\x15
```

.zshrc
```
bindkey '\e\x7F' backward-kill-word
bindkey '\ew' backward-kill-line
```

neovim config
```
if vim.fn.has("mac") == 1 then
    vim.api.nvim_set_keymap('n', '<Esc><BS>', 'db', { noremap = true, silent = true })
    vim.api.nvim_set_keymap('i', '<Esc><BS>', '<C-o>db', { noremap = true, silent = true })
end
```

**Comment by rrotter** [comment]
As an alternative to this, is there any way to get zsh, bash, and friends to bind CSI u / FixTerms?
[edit: typo]

**Reply by dkmar** [reply]
fish shell beta has support for fixterms and kitty's protocol, my bindings look like this
```
keybind = cmd+left=csi:97;5u 
keybind = cmd+right=csi:101;5u
keybind = cmd+backspace=csi:117;5u
keybind = option+backspace=csi:119;5u
```

**Reply by dkmar** [reply]
lol im switching back to ascii hex keycodes cause fzf and python3 don't interpret the CSI u.
```
keybind = cmd+left=text:\x01
keybind = cmd+right=text:\x05
keybind = cmd+backspace=text:\x15
keybind = option+backspace=text:\x17
```

**Comment by AlbatorLaho** [comment]
Somewhat related to this, `fn+cmd+backspace` (i.e. `cmd+delete`) does not delete from the cursor to end of line.
Running `cat -vte` shows that it's sending `^[[3;9~` for what that's worth...

**Reply by rrotter** [reply]
`keybind = cmd+delete=text:\x0B`
^K is the standard binding for "delete to end of line" in bash and zsh. Found what ^K is in hex [here](http://daleswanson.org/ascii.htm).
